### DISCUSSION TITLE
Can I use http protocol

### DISCUSSION THREAD

**Discussion by happy-v587** [nats-io/nats-server#2963] [General]
Can I use as `http://127.0.0.1:4222`?

**Comment by ripienaar** [comment]
no, you can use websocket though, several of our languages support that

**Reply by antoinetran** [reply]
Hi, I was wondering, even if NATS does not support http natively, but support websocket, if I setup a websocket as port 80 or 443 (with tls), then it would not be blocked by HTTP proxy? I guess that depends if the proxy does DPI?

**Reply by derekcollison** [reply]
Correct

**Reply by antoinetran** [reply]
Thanks. So I think I misread this info
```
Built to more easily traverse firewalls and load balancers, NATS WebSocket support provides even more flexibility to NATS deployments and makes it easier to communicate to the edge and endpoints.
```
(source: https://docs.nats.io/release-notes/whats_new/whats_new_22#websockets)
So I would have thought publishing a NATS server port though 443 TCP and websocket would be enough to have corporate proxy letting it. But that is wrong, it depends if:

- proxy support WebSocket
- proxy does not DPI

If I want full corporate-proxy-friendly behavior, I would need a NATS proxy that supports true HTTP (like https://github.com/sohlich/nats-proxy, which is only a POC).
After a few research, I found a topic with the same subject: https://www.reddit.com/r/NATS_io/comments/iqsvb0/rest_api_to_access_nats_messaging_system/ , and you mentionned https://resgate.io/docs/get-started/introduction/ as a REST gateway with NATS. I might try that!

**Reply by antoinetran** [reply]
Hi, can someone clarify this simple question: does NATS support WebSocket through HTTP Upgrade process? That is an important question to know if NATS is compatible with usual corporate proxy. Thank you!

**Reply by derekcollison** [reply]
For the clients that support websockets it does the HTTP upgrade as normal. Once established we run the complete NATS protocol over the websocket.
